src/category_theory/triangulated/triangulated.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2022 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+
+import category_theory.triangulated.pretriangulated
+
+/-!
+# Triangulated Categories
+
+This file contains the definition of triangulated categories, which are
+pretriangulated categories which satisfy the octahedron axiom.
+
+-/
+
+noncomputable theory
+
+namespace category_theory
+
+open limits category preadditive pretriangulated
+open_locale zero_object
+
+variables {C : Type*} [category C] [preadditive C] [has_zero_object C] [has_shift C ℤ]
+  [∀ (n : ℤ), functor.additive (shift_functor C n)] [pretriangulated C]
+
+variables {X₁ X₂ X₃ Z₁₂ Z₂₃ Z₁₃ : C} {u₁₂ : X₁ ⟶ X₂} {u₂₃ : X₂ ⟶ X₃} {u₁₃ : X₁ ⟶ X₃}
+  (comm : u₁₂ ≫ u₂₃ = u₁₃)
+  {v₁₂ : X₂ ⟶ Z₁₂} {w₁₂ : Z₁₂ ⟶ X₁⟦(1 : ℤ)⟧} (h₁₂ : triangle.mk u₁₂ v₁₂ w₁₂ ∈ dist_triang C)
+  {v₂₃ : X₃ ⟶ Z₂₃} {w₂₃ : Z₂₃ ⟶ X₂⟦(1 : ℤ)⟧} (h₂₃ : triangle.mk u₂₃ v₂₃ w₂₃ ∈ dist_triang C)
+  {v₁₃ : X₃ ⟶ Z₁₃} {w₁₃ : Z₁₃ ⟶ X₁⟦(1 : ℤ)⟧} (h₁₃ : triangle.mk u₁₃ v₁₃ w₁₃ ∈ dist_triang C)
+
+namespace triangulated
+
+include comm h₁₂ h₂₃ h₁₃
+
+/-- An octahedron is a type of datum whose existence is asserted by
+the octahedron axiom (TR 4), see https://stacks.math.columbia.edu/tag/05QK -/
+structure octahedron :=
+(m₁ : Z₁₂ ⟶ Z₁₃)
+(m₃ : Z₁₃ ⟶ Z₂₃)
+(comm₁ : v₁₂ ≫ m₁ = u₂₃ ≫ v₁₃)
+(comm₂ : w₁₂ = m₁ ≫ w₁₃)
+(comm₃ : v₁₃ ≫ m₃ = v₂₃)
+(comm₄ : w₁₃ ≫ u₁₂⟦1⟧' = m₃ ≫ w₂₃)
+(mem : triangle.mk m₁ m₃ (w₂₃ ≫ v₁₂⟦1⟧') ∈ dist_triang C)
+
+omit comm h₁₂ h₂₃ h₁₃
+
+instance (X : C) : nonempty (octahedron (comp_id (𝟙 X)) (contractible_distinguished X)
+  (contractible_distinguished X) (contractible_distinguished X)) :=
+begin
+  refine ⟨⟨0, 0, _, _, _, _, by convert contractible_distinguished (0 : C)⟩⟩,
+  all_goals { apply subsingleton.elim, },
+end
+
+namespace octahedron
+
+attribute [reassoc] comm₁ comm₃ comm₄
```

Comment:
```suggestion
attribute [reassoc] comm₁ comm₂ comm₃ comm₄
```
I think this will still be useful when rewriting backwards.

Reply:
I had tried this, but it does not work; it produces the following error message, which I do not understand:
```
invalid apply tactic, failed to unify
  (self.m₁ ≫ w₁₃) ≫ f' = (self.m₁ ≫ w₁₃) ≫ f'
with
  ?m_2 = ?m_2
```
Instead, I have just changed the direction of the equality for `comm₂`, and it works...